lib/libc/gen/syslog.3
.\" Copyright (c) 1983 Regents of the University of California.
.\" All rights reserved.  The Berkeley software License Agreement
.\" specifies the terms and conditions for redistribution.
.\"
.\"	@(#)syslog.3	6.2 (Berkeley) 09/17/85
.\"
.TH SYSLOG 3 ""
.UC 5
.SH NAME
syslog, openlog, closelog, setlogmask \- control system log
.SH SYNOPSIS
.B "#include <syslog.h>
.PP
.B "openlog(ident, logopt, facility)
.br
.B "char *ident;
.PP
.B "syslog(priority, message, parameters ... )
.br
.B "char *message;
.PP
.B "closelog()
.PP
.B "setlogmask(maskpri)
.SH DESCRIPTION
.I Syslog
arranges to write
.I message
onto the system log maintained by
.IR syslogd (8).
The message is tagged with
.IR priority .
The message looks like a
.IR printf (3)
string except that
.B %m
is replaced by the current error message (collected from
.IR errno ).
A trailing newline is added if needed.
This message will be read by
.IR syslogd (8)
and written to the system console, log files, or forwarded to
.I syslogd
on another host as appropriate.
.PP
Priorities are encoded as a
.I facility
and a
.IR level .
The facility describes the part of the system
generating the message.
The level is selected from an ordered list:
.IP LOG_EMERG \w'LOG_WARNING'u+3
A panic condition.
This is normally broadcast to all users.
.IP LOG_ALERT \w'LOG_WARNING'u+3
A condition that should be corrected immediately,
such as a corrupted system database.
.IP LOG_CRIT \w'LOG_WARNING'u+3
Critical conditions,
e.g., hard device errors.
.IP LOG_ERR \w'LOG_WARNING'u+3
Errors.
.IP LOG_WARNING \w'LOG_WARNING'u+3
Warning messages.
.IP LOG_NOTICE \w'LOG_WARNING'u+3
Conditions that are not error conditions,
but should possibly be handled specially.
.IP LOG_INFO \w'LOG_WARNING'u+3
Informational messages.
.IP LOG_DEBUG \w'LOG_WARNING'u+3
Messages that contain information
normally of use only when debugging a program.
.PP
If
.I syslog
cannot pass the message to
.I syslogd
and the message is priority LOG_ERR or higher,
it attempts to write the message on
.IR /dev/console.
.PP
If special processing is needed,
.I openlog
can be called to initialize the log file.
The parameter
.I ident
is a string that is prepended to every message.
.I Logopt
is a bit field indicating logging options.
Current values for
.I logopt
are:
.IP LOG_PID \w'LOG_WARNING'u+3
log the process id with each message:
useful for identifying instantiations of daemons.
.IP LOG_CONS \w'LOG_WARNING'u+3
Force writing messages to the console if unable to send it to
.I syslogd.
This option is safe to use in daemon processes that have no controlling
terminal since
.I syslog
will fork before opening the console.
.IP LOG_ODELAY \w'LOG_WARNING'u+3
Delay opening the connection to
.I syslogd
until the first message is logged.
Useful for programs that need to manage the
order in which file descriptors are allocated.
.PP
The
.I facility
parameter is encodes a default facility to be assigned to all messages
that do not have an explicit facility encoded:
.IP LOG_KERN \w'LOG_WARNING'u+3
Messages generated by the kernel.
These cannot be generated by any user processes.
.IP LOG_USER \w'LOG_WARNING'u+3
Messages generated by random user processes.
This is the default if none is specified.
.IP LOG_MAIL \w'LOG_WARNING'u+3
The mail system.
.IP LOG_DAEMON \w'LOG_WARNING'u+3
System daemons, such as
.IR lpd (8),
.IR routed (8),
etc.
.IP LOG_AUTH \w'LOG_WARNING'u+3
The authorization system:
.IR login (1),
.IR su (1),
.IR getty (8),
etc.
.IP LOG_LOCAL0 \w'LOG_WARNING'u+3
Reserved for local use.
Similarly for LOG_LOCAL1 through LOG_LOCAL7.
.PP
.I Closelog
can be used to close the log file.
.PP
.I Setlogmask
sets the log priority mask to
.I maskpri
and returns the previous mask.
Calls to
.I syslog
with a priority lower than
.I maskpri
are rejected.
The default allows all priorities to be logged.
.SH EXAMPLES
.nf
syslog(LOG_ALERT, "who: internal error 23");

openlog("ftpd", LOG_PID, LOG_DAEMON);
syslog(LOG_INFO, "Connection from host %d", CallingHost);

syslog(LOG_INFO|LOG_LOCAL2, "foobar error: %m");
.fi
.SH "SEE ALSO"
logger(1),
syslogd(8)


lib/libc/gen/syslog.c
/*
 * Copyright (c) 1983, 1988, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * %sccs.include.redist.c%
 */

#if defined(LIBC_SCCS) && !defined(lint)
static char sccsid[] = "@(#)syslog.c	8.2 (Berkeley) 03/17/94";
#endif /* LIBC_SCCS and not lint */

#include <sys/types.h>
#include <sys/socket.h>
#include <sys/syslog.h>
#include <sys/uio.h>
#include <netdb.h>

#include <errno.h>
#include <fcntl.h>
#include <paths.h>
#include <stdio.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#if __STDC__
#include <stdarg.h>
#else
#include <varargs.h>
#endif

static int	LogFile = -1;		/* fd for log */
static int	connected;		/* have done connect */
static int	LogStat = 0;		/* status bits, set by openlog() */
static const char *LogTag = "syslog";	/* string to tag the entry with */
static int	LogFacility = LOG_USER;	/* default facility code */
static int	LogMask = 0xff;		/* mask of priorities to be logged */

/*
 * syslog, vsyslog --
 *	print message on log file; output is intended for syslogd(8).
 */
void
#if __STDC__
syslog(int pri, const char *fmt, ...)
#else
syslog(pri, fmt, va_alist)
	int pri;
	char *fmt;
	va_dcl
#endif
{
	va_list ap;

#if __STDC__
	va_start(ap, fmt);
#else
	va_start(ap);
#endif
	vsyslog(pri, fmt, ap);
	va_end(ap);
}

void
vsyslog(pri, fmt, ap)
	int pri;
	register const char *fmt;
	va_list ap;
{
	register int cnt;
	register char *p;
	time_t now;
	int fd, saved_errno;
	char *stdp, tbuf[2048], fmt_cpy[1024];

#define	INTERNALLOG	LOG_ERR|LOG_CONS|LOG_PERROR|LOG_PID
	/* Check for invalid bits. */
	if (pri & ~(LOG_PRIMASK|LOG_FACMASK)) {
		syslog(INTERNALLOG,
		    "syslog: unknown facility/priority: %x", pri);
		pri &= LOG_PRIMASK|LOG_FACMASK;
	}

	/* Check priority against setlogmask values. */
	if (!LOG_MASK(LOG_PRI(pri)) & LogMask)
		return;

	saved_errno = errno;

	/* set default facility if none specified */
	if ((pri & LOG_FACMASK) == 0)
		pri |= LogFacility;

	/* build the message */
	(void)time(&now);
	p = tbuf + sprintf(tbuf, "<%d>", pri);
	p += strftime(p, sizeof (tbuf) - (p - tbuf), "%h %e %T ",
	    localtime(&now));
	if (LogStat & LOG_PERROR)
		stdp = p;
	if (LogTag) {
		(void)strcpy(p, LogTag);
		for (; *p; ++p);
	}
	if (LogStat & LOG_PID)
		p += sprintf(p, "[%d]", getpid());
	if (LogTag) {
		*p++ = ':';
		*p++ = ' ';
	}

	/* substitute error message for %m */
	{
		register char ch, *t1, *t2;

		for (t1 = fmt_cpy; ch = *fmt; ++fmt)
			if (ch == '%' && fmt[1] == 'm') {
				++fmt;
				for (t2 = strerror(saved_errno);
				    *t1 = *t2++; ++t1);
			}
			else
				*t1++ = ch;
		*t1 = '\0';
	}

	p += vsprintf(p, fmt_cpy, ap);
	cnt = p - tbuf;

	/* output to stderr if requested */
	if (LogStat & LOG_PERROR) {
		struct iovec iov[2];
		register struct iovec *v = iov;

		v->iov_base = stdp;
		v->iov_len = cnt - (stdp - tbuf);
		++v;
		v->iov_base = "\n";
		v->iov_len = 1;
		(void)writev(STDERR_FILENO, iov, 2);
	}

	/* get connected, output the message to the local logger */
	if (!connected)
		openlog(LogTag, LogStat | LOG_NDELAY, 0);
	if (send(LogFile, tbuf, cnt, 0) >= 0)
		return;

	/* see if should attempt the console */
	if (!(LogStat&LOG_CONS))
		return;

	/*
	 * Output the message to the console; don't worry about blocking,
	 * if console blocks everything will.  Make sure the error reported
	 * is the one from the syslogd failure.
	 */
	if ((fd = open(_PATH_CONSOLE, O_WRONLY, 0)) >= 0) {
		(void)strcat(tbuf, "\r\n");
		cnt += 2;
		p = index(tbuf, '>') + 1;
		(void)write(fd, p, cnt - (p - tbuf));
		(void)close(fd);
	}
}

static struct sockaddr SyslogAddr;	/* AF_UNIX address of local logger */

void
openlog(ident, logstat, logfac)
	const char *ident;
	int logstat, logfac;
{
	if (ident != NULL)
		LogTag = ident;
	LogStat = logstat;
	if (logfac != 0 && (logfac &~ LOG_FACMASK) == 0)
		LogFacility = logfac;

	if (LogFile == -1) {
		SyslogAddr.sa_family = AF_UNIX;
		(void)strncpy(SyslogAddr.sa_data, _PATH_LOG,
		    sizeof(SyslogAddr.sa_data));
		if (LogStat & LOG_NDELAY) {
			if ((LogFile = socket(AF_UNIX, SOCK_DGRAM, 0)) == -1)
				return;
			(void)fcntl(LogFile, F_SETFD, 1);
		}
	}
	if (LogFile != -1 && !connected)
		if (connect(LogFile, &SyslogAddr, sizeof(SyslogAddr)) == -1) {
			(void)close(LogFile);
			LogFile = -1;
		} else
			connected = 1;
}

void
closelog()
{
	(void)close(LogFile);
	LogFile = -1;
	connected = 0;
}

/* setlogmask -- set the log mask level */
int
setlogmask(pmask)
	int pmask;
{
	int omask;

	omask = LogMask;
	if (pmask != 0)
		LogMask = pmask;
	return (omask);
}
